src/avatar.cpp

```diff
@@ -1357,102 +1358,79 @@ faction *avatar::get_faction() const
-                    add_msg( _( "You set your mech's leg power to a loping fast walk." ) );
-                } else {
-                    add_msg( _( "You nudge your steed into a steady trot." ) );
-                }
-            } else {
-                add_msg( _( "You start walking." ) );
-            }
-            break;
-        }
-        case CMM_RUN: {
-            if( can_run() ) {
-                if( is_hauling() ) {
-                    stop_hauling();
-                }
-                if( is_mounted() ) {
-                    if( mounted_creature->has_flag( MF_RIDEABLE_MECH ) ) {
-                        add_msg( _( "You set the power of your mech's leg servos to maximum." ) );
-                    } else {
-                        add_msg( _( "You spur your steed into a gallop." ) );
-                    }
-                } else {
-                    add_msg( _( "You start running." ) );
-                }
-            } else {
-                if( is_mounted() ) {
-                    // mounts dont currently have stamina, but may do in future.
-                    add_msg( m_bad, _( "Your steed is too tired to go faster." ) );
-                } else if( get_working_leg_count() < 2 ) {
-                    add_msg( m_bad, _( "You need two functional legs to run." ) );
-                } else {
-                    add_msg( m_bad, _( "You're too tired to run." ) );
-                }
-                return;
-            }
-            break;
-        }
-        case CMM_CROUCH: {
-            if( is_mounted() ) {
-                if( mounted_creature->has_flag( MF_RIDEABLE_MECH ) ) {
-                    add_msg( _( "You reduce the power of your mech's leg servos to minimum." ) );
-                } else {
-                    add_msg( _( "You slow your steed to a walk." ) );
-                }
-            } else {
-                add_msg( _( "You start crouching." ) );
-            }
-            break;
-        }
-        default: {
-            return;
+    move_mode new_mode = mode.obj();
+
+    if( new_mode.minimum_required_legs > get_working_leg_count() ) {
+        add_msg( m_bad, string_format( _( "You do not have enough functioning legs to %s." ),
+                                       new_mode.name ) );
+        return false;
+    }
+
+    if( !new_mode.usable_while_mounted && is_mounted() ) {
+        add_msg( m_bad, string_format( _( "You cannot %s whilst mounted." ), new_mode.name ) );
+        return false;
+    }
+
+    if( !new_mode.can_haul && is_hauling() ) {
+        stop_hauling();
+    }
+
+    translation text_to_show = new_mode.flavor_text;
+    if( is_mounted() ) {
+        if( mounted_creature->has_flag( MF_RIDEABLE_MECH ) ) {
+            text_to_show = new_mode.flavor_text_mech;
+        } else {
+            text_to_show = new_mode.flavor_text_mount;
         }
     }
-    move_mode = new_mode;
+
+    add_msg( text_to_show.translated() );
```

Comment:
```suggestion
    add_msg( "%s", text_to_show.translated() );
```
Otherwise you'll get a problem when the translated text contains a "%". `add_msg` will think this has to be replaced with the content of another argument, but there are no further arguments.